packages/blocks/src/_specs/preset/mobile-patch.ts

```diff
@@ -0,0 +1,103 @@
+import type { Container } from '@blocksuite/global/di';
+
+import {
+  type ReferenceNodeConfig,
+  ReferenceNodeConfigIdentifier,
+} from '@blocksuite/affine-components/rich-text';
+import {
+  type BlockStdScope,
+  LifeCycleWatcher,
+  WidgetViewMapIdentifier,
+  type WidgetViewMapType,
+} from '@blocksuite/block-std';
+
+import { pageRootWidgetViewMap } from '../../root-block/page/page-root-spec.js';
+import { AFFINE_EMBED_CARD_TOOLBAR_WIDGET } from '../../root-block/widgets/embed-card-toolbar/embed-card-toolbar.js';
+import { AFFINE_FORMAT_BAR_WIDGET } from '../../root-block/widgets/format-bar/format-bar.js';
+import { AFFINE_SLASH_MENU_WIDGET } from '../../root-block/widgets/slash-menu/index.js';
+
+export class MobileSpecsPatches extends LifeCycleWatcher {
+  static override key = 'mobile-patches';
+
+  constructor(std: BlockStdScope) {
+    super(std);
+
+    std.doc.awarenessStore.setFlag('enable_mobile_keyboard_toolbar', true);
```

Comment:
Why we need to set flag here🤔

Reply:
It may be better to add both widget to root widgets map.

Reply:
After trying, it is difficult to control component rendering through extension patch. These two `setFlag` lines are a good example:

- `keyboard-toolbar` is an independent widget for the mobile version, and it is easy to control by overriding the `pageRootWidgetViewMap`.
- However, the latter flag is used to control whether to enable the mobile at-menu, and the mobile at-menu shares most of its logic with the desktop version. So, it is difficult to achieve this goal using the spec patch method only.

```ts
// affine-linked-doc-widget

  show = (mode: 'desktop' | 'mobile' = 'desktop') => {
    // ...

    const enableMobile = this.doc.awarenessStore.getFlag(
      'enable_mobile_linked_doc_menu'
    );

    this._show$.value = enableMobile ? mode : 'desktop';
  };

  override render() {
    if (this._show$.value === 'none') return nothing;

    return html`${this._renderInputMask()}
      <blocksuite-portal
        .shadowDom=${false}
        .template=${choose(
          this._show$.value,
          [
            ['desktop', this._renderLinkedDocPopover],
            ['mobile', this._renderLinkedDocMenu],
          ],
          () => html`${nothing}`
        )}
      ></blocksuite-portal>`;
  }
```

In conclusion, I think setting both flags in the constructor of the patch extension is acceptable. They can be removed in the next version release.